File: projects/x402-api/sdk-examples/go/complete_example.go
package main

import (
	"bytes"
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strconv"
	"sync"
	"time"
)

// ==================== 数据结构 ====================

// APIResponse API 响应结构
type APIResponse struct {
	Success bool        `json:"success"`
	Data    interface{} `json:"data"`
	Cost    float64     `json:"cost"`
	Error   string      `json:"error,omitempty"`
}

// TranslateRequest 翻译请求
type TranslateRequest struct {
	Text   string `json:"text"`
	Source string `json:"source"`
	Target string `json:"target"`
}

// TranslateData 翻译结果数据
type TranslateData struct {
	TranslatedText string `json:"translated_text"`
	SourceLanguage string `json:"source_language"`
	TargetLanguage string `json:"target_language"`
}

// x402Client x402 API 客户端
type x402Client struct {
	walletKey  string
	baseURL    string
	balance    float64
	totalSpent float64
	mu         sync.Mutex
}

// NewClient 创建新客户端
func NewClient(walletKey, baseURL string) *x402Client {
	if baseURL == "" {
		baseURL = "http://localhost:5002"
	}
	return &x402Client{
		walletKey:  walletKey,
		baseURL:    baseURL,
		balance:    10.0, // 模拟初始余额
		totalSpent: 0.0,
	}
}

// generateSignature 生成支付签名
func (c *x402Client) generateSignature(amount float64, endpoint string, timestamp int64) string {
	data := fmt.Sprintf("%.6f:%s:%d", amount, endpoint, timestamp)
	h := hmac.New(sha256.New, []byte(c.walletKey))
	h.Write([]byte(data))
	return hex.EncodeToString(h.Sum(nil))
}

// callAPI 通用 API 调用方法
func (c *x402Client) callAPI(endpoint string, payload interface{}, cost float64) (*APIResponse, error) {
	c.mu.Lock()
	defer c.mu.Unlock()

	// 检查余额
	if c.balance < cost {
		return nil, fmt.Errorf("余额不足：需要 %.2f USDC，当前 %.2f USDC", cost, c.balance)
	}

	// 准备请求
	timestamp := time.Now().UnixNano() / 1e6 // 毫秒时间戳
	signature := c.generateSignature(cost, endpoint, timestamp)

	jsonData, err := json.Marshal(payload)
	if err != nil {
		return nil, err
	}

	req, err := http.NewRequest("POST", c.baseURL+"/api/v1/"+endpoint, bytes.NewBuffer(jsonData))
	if err != nil {
		return nil, err
	}

	// 设置支付头
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("X-Payment-Amount", strconv.FormatInt(int64(cost*1000000), 10))
	req.Header.Set("X-Payment-Token", "USDC")
	req.Header.Set("X-Payment-Signature", signature)
	req.Header.Set("X-Payment-Timestamp", strconv.FormatInt(timestamp, 10))

	// 发送请求
	client := &http.Client{Timeout: 30 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("API 错误：%d %s", resp.StatusCode, resp.Status)
	}

	var result APIResponse
	if err := json.Unmarshal(body, &result); err != nil {
		return nil, err
	}

	// 扣减余额
	c.balance -= cost
	c.totalSpent += cost
	result.Cost = cost

	return &result, nil
}

// Translate 翻译 API
func (c *x402Client) Translate(text, source, target string) (*APIResponse, error) {
	payload := TranslateRequest{
		Text:   text,
		Source: source,
		Target: target,
	}
	return c.callAPI("translator", payload, 0.02)
}

// CodeGen 代码生成 API
func (c *x402Client) CodeGen(prompt, language string) (*APIResponse, error) {
	payload := map[string]string{
		"prompt":   prompt,
		"language": language,
	}
	return c.callAPI("code-gen", payload, 0.08)
}

// BatchTranslate 批量翻译（享受折扣）
func (c *x402Client) BatchTranslate(texts []TranslateRequest, discount float64) ([]*APIResponse, error) {
	results := make([]*APIResponse, 0, len(texts))
	costPerCall := 0.02 * discount

	for _, task := range texts {
		result, err := c.callAPI("translator", task, costPerCall)
		if err != nil {
			results = append(results, &APIResponse{
				Success: false,
				Error:   err.Error(),
				Cost:    0,
			})
		} else {
			results = append(results, result)
		}
	}

	return results, nil
}

// GetUsageStats 获取使用统计
func (c *x402Client) GetUsageStats() map[string]interface{} {
	c.mu.Lock()
	defer c.mu.Unlock()

	usageRate := (c.totalSpent / 10.0) * 100

	return map[string]interface{}{
		"initial_balance": 10.0,
		"current_balance": c.balance,
		"total_spent":     c.totalSpent,
		"usage_rate":      fmt.Sprintf("%.1f%%", usageRate),
	}
}

// ==================== 使用示例 ====================

func exampleBasicUsage() {
	fmt.Println("=" + string(make([]byte, 50)))
	fmt.Println("示例 1: 基础调用")
	fmt.Println("=" + string(make([]byte, 50)))

	client := NewClient("your_wallet_key", "")

	result, err := client.Translate("Hello, world!", "en", "zh")
	if err != nil {
		fmt.Printf("❌ 错误：%v\n", err)
	} else {
		data := result.Data.(map[string]interface{})
		fmt.Printf("✅ 翻译成功：%v\n", data["translated_text"])
		fmt.Printf("💰 花费：%.2f USDC\n\n", result.Cost)
	}
}

func exampleErrorHandling() {
	fmt.Println("=" + string(make([]byte, 50)))
	fmt.Println("示例 2: 错误处理")
	fmt.Println("=" + string(make([]byte, 50)))

	client := NewClient("your_wallet_key", "")
	client.balance = 0.01 // 余额不足

	_, err := client.Translate("Hello", "en", "zh")
	if err != nil {
		fmt.Printf("⚠️ 错误：%v\n", err)
		fmt.Println("💡 建议：检查余额或充值\n")
	}
}

func exampleBatchProcessing() {
	fmt.Println("=" + string(make([]byte, 50)))
	fmt.Println("示例 3: 批量调用（享受折扣）")
	fmt.Println("=" + string(make([]byte, 50)))

	client := NewClient("your_wallet_key", "")

	// 准备 15 条翻译任务
	texts := make([]TranslateRequest, 15)
	for i := 0; i < 15; i++ {
		texts[i] = TranslateRequest{
			Text:   fmt.Sprintf("Message %d", i),
			Source: "en",
			Target: "zh",
		}
	}

	results, err := client.BatchTranslate(texts, 0.8) // 8 折
	if err != nil {
		fmt.Printf("❌ 错误：%v\n", err)
		return
	}

	totalCost := 0.0
	successful := 0
	for _, r := range results {
		if r.Success {
			totalCost += r.Cost
			successful++
		}
	}

	fmt.Printf("📊 批量翻译 %d/%d 条\n", successful, len(texts))
	fmt.Printf("💰 总花费：%.4f USDC\n", totalCost)
	fmt.Printf("🎯 平均每次：%.4f USDC\n", totalCost/float64(successful))
	fmt.Printf("💡 节省了：%.4f USDC\n\n", 0.02*15-totalCost)
}

func exampleConcurrentCalls() {
	fmt.Println("=" + string(make([]byte, 50)))
	fmt.Println("示例 4: 并发调用")
	fmt.Println("=" + string(make([]byte, 50)))

	client := NewClient("your_wallet_key", "")

	// 并发执行
	var wg sync.WaitGroup
	results := make(chan *APIResponse, 5)

	for i := 0; i < 5; i++ {
		wg.Add(1)
		go func(n int) {
			defer wg.Done()
			result, _ := client.Translate(fmt.Sprintf("Text %d", n), "en", "zh")
			results <- result
		}(i)
	}

	wg.Wait()
	close(results)

	successful := 0
	totalCost := 0.0
	for r := range results {
		if r.Success {
			successful++
			totalCost += r.Cost
		}
	}

	fmt.Printf("✅ 成功：%d/%d\n", successful, 5)
	fmt.Printf("💰 总花费：%.4f USDC\n", totalCost)
	fmt.Printf("⚡ 并发执行，节省时间\n\n")
}

func exampleUsageMonitoring() {
	fmt.Println("=" + string(make([]byte, 50)))
	fmt.Println("示例 5: 使用量监控")
	fmt.Println("=" + string(make([]byte, 50)))

	client := NewClient("your_wallet_key", "")

	// 模拟使用
	for i := 0; i < 10; i++ {
		client.Translate(fmt.Sprintf("Test %d", i), "en", "zh")
	}

	stats := client.GetUsageStats()

	fmt.Println("📊 使用统计:")
	fmt.Printf("  初始余额：$%.2f USDC\n", stats["initial_balance"])
	fmt.Printf("  当前余额：$%.2f USDC\n", stats["current_balance"])
	fmt.Printf("  总花费：$%.2f USDC\n", stats["total_spent"])
	fmt.Printf("  使用率：%s\n", stats["usage_rate"])

	if stats["current_balance"].(float64) < 2.0 {
		fmt.Println("\n⚠️ 警告：余额低于 $2 USDC，请充值！")
	}

	fmt.Println()
}

func exampleRealWorldApp() {
	fmt.Println("=" + string(make([]byte, 50)))
	fmt.Println("示例 6: 真实应用 - 多语言客服系统")
	fmt.Println("=" + string(make([]byte, 50)))

	client := NewClient("your_wallet_key", "")

	conversations := []struct {
		User string
		Lang string
	}{
		{"Hello, I need help", "en"},
		{"Bonjour, j'ai un problème", "fr"},
		{"Hola, necesito ayuda", "es"},
	}

	fmt.Println("🎧 客服系统启动...\n")

	for _, conv := range conversations {
		result, err := client.Translate(conv.User, conv.Lang, "zh")
		if err != nil {
			fmt.Printf("❌ 错误：%v\n", err)
			continue
		}

		if result.Success {
			data := result.Data.(map[string]interface{})
			fmt.Printf("[%s] 用户：%s\n", conv.Lang, conv.User)
			fmt.Printf("[zh] 翻译：%v\n", data["translated_text"])
			fmt.Printf("💰 花费：%.2f USDC\n\n", result.Cost)
		}
	}

	fmt.Printf("💵 总成本：$%.2f USDC\n", client.totalSpent)
	fmt.Println("💡 如果自建翻译团队：约 $50/小时")
	fmt.Printf("🎯 使用 x402 API: $%.2f USDC\n", client.totalSpent)
	fmt.Printf("✅ 节省了：$%.2f USDC\n\n", 50-client.totalSpent)
}

func main() {
	fmt.Println("\n🚀 x402 API Go SDK 完整示例\n")

	exampleBasicUsage()
	exampleErrorHandling()
	exampleBatchProcessing()
	exampleConcurrentCalls()
	exampleUsageMonitoring()
	exampleRealWorldApp()

	fmt.Println("=" + string(make([]byte, 50)))
	fmt.Println("✅ 所有示例运行完成！")
	fmt.Println("=" + string(make([]byte, 50)))
}

File: projects/x402-api/sdk-examples/python/complete_example.py
#!/usr/bin/env python3
"""
x402 API - Python SDK 完整示例

包含：
- 基础调用
- 错误处理
- 批量调用
- 异步调用
- 使用量监控
"""

import asyncio
from typing import List, Dict
from dataclasses import dataclass

# 模拟 SDK（实际使用时 pip install x402-sdk）
@dataclass
class APIResponse:
    success: bool
    data: Dict
    cost: float
    error: str = None

class x402Client:
    """x402 API 客户端示例"""
    
    def __init__(self, wallet_key: str, base_url: str = "http://8.213.149.224:5002"):
        self.wallet_key = wallet_key
        self.base_url = base_url
        self.balance = 10.0  # 模拟余额
        self.total_spent = 0.0
    
    async def translate(self, text: str, source: str, target: str) -> APIResponse:
        """翻译 API"""
        cost = 0.02
        
        if self.balance < cost:
            return APIResponse(success=False, data={}, cost=0, error="余额不足")
        
        # 模拟 API 调用
        await asyncio.sleep(0.1)
        
        self.balance -= cost
        self.total_spent += cost
        
        return APIResponse(
            success=True,
            data={"translated_text": f"[{source}→{target}] {text}"},
            cost=cost
        )
    
    async def code_gen(self, prompt: str, language: str = "python") -> APIResponse:
        """代码生成 API"""
        cost = 0.08
        
        if self.balance < cost:
            return APIResponse(success=False, data={}, cost=0, error="余额不足")
        
        await asyncio.sleep(0.2)
        
        self.balance -= cost
        self.total_spent += cost
        
        return APIResponse(
            success=True,
            data={"code": f"# {language} code for: {prompt}\ndef solution():\n    pass"},
            cost=cost
        )
    
    async def batch_translate(self, texts: List[Dict]) -> List[APIResponse]:
        """批量翻译"""
        results = []
        
        # 批量折扣：超过 10 次打 8 折
        discount = 0.8 if len(texts) > 10 else 1.0
        
        for task in texts:
            cost = 0.02 * discount
            result = await self.translate(
                text=task['text'],
                source=task.get('source', 'en'),
                target=task.get('target', 'zh')
            )
            result.cost = cost
            results.append(result)
        
        return results
    
    def get_usage_stats(self) -> Dict:
        """获取使用统计"""
        return {
            "initial_balance": 10.0,
            "current_balance": self.balance,
            "total_spent": self.total_spent,
            "usage_rate": f"{(self.total_spent / 10.0 * 100):.1f}%"
        }


# ==================== 使用示例 ====================

async def example_basic_usage():
    """基础使用示例"""
    print("=" * 50)
    print("示例 1: 基础调用")
    print("=" * 50)
    
    client = x402Client(wallet_key="your_wallet_key")
    
    # 单次调用
    result = await client.translate("Hello, world!", "en", "zh")
    
    if result.success:
        print(f"✅ 翻译成功：{result.data['translated_text']}")
        print(f"💰 花费：{result.cost} USDC")
    else:
        print(f"❌ 失败：{result.error}")
    
    print()


async def example_error_handling():
    """错误处理示例"""
    print("=" * 50)
    print("示例 2: 错误处理")
    print("=" * 50)
    
    client = x402Client(wallet_key="your_wallet_key")
    client.balance = 0.01  # 余额不足
    
    result = await client.translate("Hello", "en", "zh")
    
    if not result.success:
        print(f"⚠️ 错误：{result.error}")
        print("💡 建议：检查余额或充值")
    
    print()


async def example_batch_processing():
    """批量处理示例"""
    print("=" * 50)
    print("示例 3: 批量调用（享受折扣）")
    print("=" * 50)
    
    client = x402Client(wallet_key="your_wallet_key")
    
    texts = [
        {"text": f"Message {i}", "source": "en", "target": "zh"}
        for i in range(15)  # 15 次，享受 8 折
    ]
    
    results = await client.batch_translate(texts)
    
    total_cost = sum(r.cost for r in results)
    print(f"📊 批量翻译 {len(results)} 条")
    print(f"💰 总花费：{total_cost:.4f} USDC")
    print(f"🎯 平均每次：{total_cost/len(results):.4f} USDC")
    print(f"💡 节省了：{(0.02*15 - total_cost):.4f} USDC")
    
    print()


async def example_concurrent_calls():
    """并发调用示例"""
    print("=" * 50)
    print("示例 4: 并发调用")
    print("=" * 50)
    
    client = x402Client(wallet_key="your_wallet_key")
    
    # 并发执行多个任务
    tasks = [
        client.translate(f"Text {i}", "en", "zh")
        for i in range(5)
    ]
    
    results = await asyncio.gather(*tasks)
    
    successful = sum(1 for r in results if r.success)
    total_cost = sum(r.cost for r in results if r.success)
    
    print(f"✅ 成功：{successful}/{len(results)}")
    print(f"💰 总花费：{total_cost:.4f} USDC")
    print(f"⚡ 并发执行，节省时间")
    
    print()


async def example_usage_monitoring():
    """使用量监控示例"""
    print("=" * 50)
    print("示例 5: 使用量监控")
    print("=" * 50)
    
    client = x402Client(wallet_key="your_wallet_key")
    
    # 模拟使用
    for i in range(10):
        await client.translate(f"Test {i}", "en", "zh")
    
    # 查看统计
    stats = client.get_usage_stats()
    
    print(f"📊 使用统计:")
    print(f"  初始余额：${stats['initial_balance']} USDC")
    print(f"  当前余额：${stats['current_balance']:.2f} USDC")
    print(f"  总花费：${stats['total_spent']:.2f} USDC")
    print(f"  使用率：{stats['usage_rate']}")
    
    # 告警
    if stats['current_balance'] < 2.0:
        print(f"\n⚠️ 警告：余额低于 $2 USDC，请充值！")
    
    print()


async def example_real_world_app():
    """真实应用场景：多语言客服系统"""
    print("=" * 50)
    print("示例 6: 真实应用 - 多语言客服系统")
    print("=" * 50)
    
    client = x402Client(wallet_key="your_wallet_key")
    
    # 模拟客服对话
    conversations = [
        {"user": "Hello, I need help", "lang": "en"},
        {"user": "Bonjour, j'ai un problème", "lang": "fr"},
        {"user": "Hola, necesito ayuda", "lang": "es"},
    ]
    
    print("🎧 客服系统启动...\n")
    
    for conv in conversations:
        # 翻译用户消息到中文
        result = await client.translate(
            conv['user'],
            conv['lang'],
            'zh'
        )
        
        if result.success:
            print(f"[{conv['lang']}] 用户：{conv['user']}")
            print(f"[zh] 翻译：{result.data['translated_text']}")
            print(f"💰 花费：{result.cost} USDC\n")
    
    print(f"💵 总成本：${client.total_spent:.2f} USDC")
    print(f"💡 如果自建翻译团队：约 $50/小时")
    print(f"🎯 使用 x402 API：${client.total_spent:.2f} USDC")
    print(f"✅ 节省了：${50 - client.total_spent:.2f} USDC")
    
    print()


async def main():
    """运行所有示例"""
    print("\n🚀 x402 API Python SDK 完整示例\n")
    
    await example_basic_usage()
    await example_error_handling()
    await example_batch_processing()
    await example_concurrent_calls()
    await example_usage_monitoring()
    await example_real_world_app()
    
    print("=" * 50)
    print("✅ 所有示例运行完成！")
    print("=" * 50)


if __name__ == "__main__":
    asyncio.run(main())
